olp-cpp-sdk-core/include/olp/core/thread/TaskContinuation.h

```diff
@@ -0,0 +1,397 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <deque>
+#include <functional>
+#include <memory>
+
+#include <olp/core/client/ApiError.h>
+#include <olp/core/client/ApiResponse.h>
+#include <olp/core/client/TaskContext.h>
+#include <olp/core/porting/make_unique.h>
+
+namespace olp {
+namespace thread {
+
+class TaskScheduler;
+
+template <typename ResultType>
+class Continuation;
+
+namespace internal {
+
+template <typename Type>
+struct RemoveRefAndConstImpl {
+  using type =
+      typename std::remove_cv<typename std::remove_reference<Type>::type>::type;
+};
+
+template <typename Type>
+using RemoveRefAndConst = typename RemoveRefAndConstImpl<Type>::type;
+
+template <typename Type>
+struct ReduceVoidVectorImpl {
+  using type =
+      typename std::conditional<std::is_same<Type, std::vector<void>>::value,
+                                void, Type>::type;
+};
+
+template <typename Type>
+using ReduceVoidVector = typename ReduceVoidVectorImpl<Type>::type;
+
+template <typename Type>
+struct TypeToFunctionInputImpl {
+  using type = std::function<void(Type)>;
+};
+
+template <>
+struct TypeToFunctionInputImpl<void> {
+  using type = std::function<void()>;
+};
+
+template <typename Type>
+using TypeToFunctionInput = typename TypeToFunctionInputImpl<Type>::type;
+
+template <typename Class, typename ExecutionContext, typename Arg>
+Arg secondArgType(void (Class::*)(ExecutionContext, Arg) const);
+
+template <typename Class, typename ExecutionContext, typename Arg>
+Arg secondArgType(void (Class::*)(ExecutionContext, Arg));
+
+template <typename Class, typename Some, typename ExecutionContext,
+          typename Arg>
+Arg thirdArgType(void (Class::*)(Some, ExecutionContext, Arg) const);
+
+template <typename Class, typename Some, typename ExecutionContext,
+          typename Arg>
+Arg thirdArgType(void (Class::*)(Some, ExecutionContext, Arg));
+
+void funcArgType(std::function<void()>);
+
+template <typename Arg>
+Arg funcArgType(std::function<void(Arg)>);
+
+template <typename Callable>
+struct AsyncResultTypeImpl {
+  using FirstArgType =
+      RemoveRefAndConst<decltype(secondArgType(&Callable::operator()))>;
+  using type = RemoveRefAndConst<decltype(funcArgType(FirstArgType()))>;
+};
+
+template <typename Callable>
+using AsyncResultType = typename AsyncResultTypeImpl<Callable>::type;
+
+template <typename Callable>
+struct AsyncResultTypeWithExecutionContextFullImpl {
+  using ThirdArgType =
+      RemoveRefAndConst<decltype(thirdArgType(&Callable::operator()))>;
+  using type = RemoveRefAndConst<decltype(funcArgType(ThirdArgType()))>;
+};
+
+template <typename Callable>
+using AsyncResultExecutionContextTypeFull =
```

Comment:
I would consider a shorter name, as I see it will become very big in the below Continuation class.

Reply:
renamed to DeducedType